shotgun_api3/shotgun.py

```diff
@@ -2155,6 +2156,132 @@ def upload(self, entity_type, entity_id, path, field_name=None, display_name=Non
         is_thumbnail = (field_name in ["thumb_image", "filmstrip_thumb_image", "image",
                                        "filmstrip_image"])
 
+        # Thumbnails can only be uploaded to SG for now...
+        if self.server_info['s3_direct_uploads_enabled'] \
+                and entity_type == 'Version' and field_name == 'sg_uploaded_movie':
+            return self._upload_to_storage(entity_type, entity_id, path, field_name, display_name,
+                                           tag_list, is_thumbnail)
+        else:
+            return self._upload_to_sg(entity_type, entity_id, path, field_name, display_name,
+                                      tag_list, is_thumbnail)
+
+    def _upload_to_storage(self, entity_type, entity_id, path, field_name, display_name,
+                           tag_list, is_thumbnail):
+        """
+        Internal function to upload a file to S3 and link it to the specified entity.
+
+        :param str entity_type: Entity type to link the upload to.
+        :param int entity_id: Id of the entity to link the upload to.
+        :param str path: Full path to an existing non-empty file on disk to upload.
+        :param str field_name: The internal Shotgun field name on the entity to store the file in.
+            This field must be a File/Link field type.
+        :param str display_name: The display name to use for the file. Defaults to the file name.
+        :param str tag_list: comma-separated string of tags to assign to the file.
+        :param bool is_thumbnail: indicates if the attachment is a thumbnail.
+        :returns: Id of the Attachment entity that was created for the image.
+        :rtype: int
+        """
+
+        filename = os.path.basename(path)
+
+        # Step 1: get the upload url
+
+        upload_info = self._get_attachment_upload_info(is_thumbnail, filename)
+
+        # Step 2: upload the file
+
+        # Get some file info
+        fd = open(path, "rb")
+        content_type = mimetypes.guess_type(filename)[0]
+        content_type = content_type or "application/octet-stream"
+        file_size = os.fstat(fd.fileno())[stat.ST_SIZE]
+
+        # Perform the request
+        opener = urllib2.build_opener(urllib2.HTTPHandler)
+        try:
+            file = open(path, "rb")
```

Comment:
```
fd = open(path, "rb")
try:
    ...
except urllib2.HTTPError, e:
    ...
finally:
    fd.close()
```
(also note - nitpicky - that file is a reserved word in python so better to use some other variable name)

So if we use `os.path.getsize(path)` as suggested above, we don't need to open the file until we get to this point in the code.